src/plugins/procdump/procdump.cpp

```diff
@@ -0,0 +1,777 @@
+    }
+    else
+    {
+        PRINT_DEBUG("[PROCDUMP] [PID:%d] Error: Dump %u PTEs from %u / %lu\n",
+                    ctx->pid, ptes_to_dump, idx, total_number_of_ptes);
+        g_assert(false);
+    }
+
+    return RTLCOPY_INJECT;
+}
+
+static bool dump_next_vads(drakvuf_t drakvuf, drakvuf_trap_info_t* info,
+                           procdump_ctx* ctx)
+{
+    while (!ctx->vads.empty())
+    {
+        if (2 == ctx->vads.begin()->second.type)
+        {
+            switch(dump_with_rtlcopymemory(drakvuf, info, ctx))
+            {
+            case RTLCOPY_INJECT:
+                return true;
+            case RTLCOPY_GO_NEXT_VAD:
+                continue;
+            case RTLCOPY_RETRY_WITH_MMAP:
+            default:
+                break;
+            }
+        }
+
+        dump_with_mmap(drakvuf, info, ctx);
+    }
+
+    return false;
+}
+
+static event_response_t detach(drakvuf_t drakvuf, drakvuf_trap_info_t* info,
+                               procdump_ctx* ctx)
+{
+    if (ctx->vads.empty())
+    {
+        // If there is no VADs left than the file have been processed
+        save_file_metadata(ctx, &info->proc_data);
+        switch (ctx->plugin->m_output_format)
+        {
+            case OUTPUT_KV:
+                printf("procdump Time=" FORMAT_TIMEVAL
+                       ",PID=%d,PPID=%d,ProcessName=\"%s\",DumpReason="
+                       "\"TerminateProcess\",DumpSize=%" PRIu64 ",SN=%lu\n",
+                       UNPACK_TIMEVAL(info->timestamp), ctx->pid, ctx->ppid,
+                       ctx->name.data(), ctx->size, ctx->idx);
+                break;
+            default:
+                break;
+        }
+    }
+    // One could not restore all registers at once like this:
+    //     memcpy(info->regs, &injector->saved_regs, sizeof(x86_registers_t)),
+    // because thus kernel structures could be affected.
+    // For example on Windows 7 x64 GS BASE stores pointer to KPCR. If save
+    // GS BASE on vCPU0 and start injections Windows scheduler could switch
+    // thread to other vCPU1. After restoring all registers vCPU1's GS BASE
+    // would point to KPCR of vCPU0.
+    info->regs->rax = ctx->saved_regs.rax;
+    info->regs->rcx = ctx->saved_regs.rcx;
+    info->regs->rdx = ctx->saved_regs.rdx;
+    info->regs->rbx = ctx->saved_regs.rbx;
+    info->regs->rbp = ctx->saved_regs.rbp;
+    info->regs->rsp = ctx->saved_regs.rsp;
+    info->regs->rdi = ctx->saved_regs.rdi;
+    info->regs->rsi = ctx->saved_regs.rsi;
+    info->regs->r8 = ctx->saved_regs.r8;
+    info->regs->r9 = ctx->saved_regs.r9;
+    info->regs->r10 = ctx->saved_regs.r10;
+    info->regs->r11 = ctx->saved_regs.r11;
+    info->regs->r12 = ctx->saved_regs.r12;
+    info->regs->r13 = ctx->saved_regs.r13;
+    info->regs->r14 = ctx->saved_regs.r14;
+    info->regs->r15 = ctx->saved_regs.r15;
+
+    free_pool(ctx->plugin->pools, ctx->pool);
+    drakvuf_remove_trap(drakvuf, ctx->bp, (drakvuf_trap_free_t)g_free);
```

Comment:
Will need a wrapper function to pass for the free routine that calls `g_slice_free`

Reply:
Done